app/utils/hash_utils.rb

```diff
@@ -144,4 +144,39 @@ def flatten(h)
 
     acc.call(nil, h)
   end
+
+  # Recursively walks through nested hash and performs `map` operation.
+  #
+  # The tree is traversed in pre-order manner.
+  #
+  # In each node, calls the block with two arguments: key and value.
+  # The block needs to return a tuple of [key, value].
+  #
+  # Example (double all values):
+  #
+  # deep_map(a: { b: { c: 1}, d: [{ e: 1, f: 2 }]}) { |k, v|
+  #   [k, v * 2]
+  # }
+  #
+  #
+  # Example (stringify keys):
+  #
+  # deep_map(a: 1, b: 2) { |k, v|
+  #   [k.to_s, v]
+  # }
+  #
+  # Unlike Ruby's Hash#map, this method returns a Hash, not an Array.
+  #
+  def deep_map(obj, &block)
```

Comment:
I'm again thinking about naming and what's the general case here. This is similar to tree walking except that in a typical tree walk you can process the non-leaf nodes too if you want to. This impl would be something like `prewalk-leaves`, which really is just a special case of `prewalk`. See e.g. https://clojure.github.io/clojure/clojure.walk-api.html